src/QsCompiler/Compiler/CompilationLoader.cs

```diff
@@ -837,7 +837,7 @@ private QsCompilation ReplaceTargetSpecificImplementations(IEnumerable<string> p
                 LogError(ErrorCode.ConflictsInTargetSpecificDecompositions, Array.Empty<string>());
             }
 
-            var targetSpecificDecompositions = new QsCompilation(replacements, ImmutableArray<QsQualifiedName>.Empty);
+            var targetSpecificDecompositions = new QsCompilation(replacements, ImmutableArray<QsQualifiedName>.Empty, QsNullable<ICallGraph>.Null);

```

Comment:
I'm wondering if it's possible for QsCompilation to automatically build a call graph when it is constructed, if one isn't provided? It would be convenient if the call graph wasn't behind a QsNullable, are there cases where it is actually necessary to not build the call graph?

Reply:
When a call graph isn't provided, it is usually because one is not needed or wanted in that circumstance. For example, in the CompilationLoader's SerializeSyntaxTree method, we don't want to serialize a call graph along with the syntax tree, so we want a compilation that doesn't have a call graph.

Reply:
If we want to serialize the compilation without the call graph, I feel like that should be a separate data structure that only has the fields we want to serialize, rather than always serializing a null value. Are there other instances besides serialization where a null call graph is needed?

Reply:
To explain a little more, my main concern is for rewrite steps, transformations, etc. that work on the compilation, is the call graph always going to be present? If so, it would be more convenient to remove the nullability from it so they do not need to handle the case where the call graph is not there.

Reply:
By the time we are doing rewrite steps, the main compilation that gets pass from one rewrite step to the next always has a call graph. But there are some instances where secondary compilations are made that don't need the call graph, such as when we are merging in intrinsic decomposition.

Reply:
What happens when a rewrite step changes the compilation but doesn't update the call graph? Does the next rewrite step get an out-of-date call graph?

Since the call graph is

1. derived entirely from the other fields in the compilation
2. often invalidated/out-of-date
2. not always needed

I am sort of thinking it might make sense to have it be a computed property instead. Perhaps using `lazy` or [`System.Lazy`](https://docs.microsoft.com/en-us/dotnet/api/system.lazy-1?view=netcore-3.1), so the call graph is always up to date and always computed when needed, but never computed when it isn't needed?

Reply:
I've updated the docs PR for the call graph with the solution to out-of-date call graphs. It is a strategy that involves comparing guids. If you'd like to discuss a different strategy, like your suggestion of using a lazy evaluation, we can do that offline. But either way that would be part of a different PR.

Reply:
On second though, the `lazy` call graph idea might be easy enough to try out on this PR. Let me do some looking into it and get back to you.

Reply:
For this PR, my concern is primarily about QsNullable. It seems like the cases where the call graph is present outnumber the cases where it isn't - let me know if I misunderstood that. I am interested in making the common case convenient and type-safe, but with a QsNullable I am imagining that rewrite steps will either:

1. Assume that the call graph is always there and unsafely access QsNullable.Item. If the rewrite step is used in a way that means the call graph isn't there, it will crash.
2. Check if the call graph is there and handle both cases, potentially skipping the rewrite step or manually building the call graph if it is missing. However, if the call graph is always there, every rewrite step will need to do unnecessary checks, adding to the required boilerplate for rewrite steps that use the call graph

Please let me know if my assessment of the situation isn't accurate. :) Otherwise, is there a way to resolve this in this PR?

Reply:
Yah most of the time the main compilation will have a non-null call graph. I'm still looking into the `lazy` call graph approach to see if it'll work the way we need it to. One practical concern I have with that approach is that the QsCompiler object will need to have the ability to create call graphs, which means it would need access to the call graph walker. This might create a dependency issue.

Reply:
Hmm, yeah, this circular dependency between the call graph and the compilation is an issue...

What are the disadvantages to building it on-the-fly instead of storing it in `QsCompilation`? For performance, the compilation manager could still cache it, but if it wasn't stored, rewrite steps never need to worry about accidentally having an out-of-date call graph, and things that create a `QsCompilation` but don't care about the call graph don't need to do unnecessary work...

The one case where this would cause duplicate work that I can think of is:

1. A `QsCompilation` is passed from one place to another (so they couldn't share a cache),
2. without being modified (so the call graph wouldn't be out-of-date), and
3. both places need the call graph?

Is that common?

Reply:
With how often the call graph is going to be made out of date (every time a new QsCompilation is made), maybe it does make sense to not have it as a field on the compilation, and just have all the locations create one on-the-fly as needed. This would certainly simplify the changes to the QsCompilation (eliminate the changes, actually). Would we still want to keep around the interfaces for the call graph classes?

Reply:
I think if the call graph isn't in `QsCompilation` (or some other type in the DataStructures project), then that probably removes the need to have the interfaces.

Reply:
As per this discussion, the call graph field of the compilation has been removed in favor of the design where, if a piece of code needs a call graph, they generate it. The interfaces have been removed as they are no longer necessary.